core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/api/rename.kt

```diff
@@ -0,0 +1,46 @@
+package org.jetbrains.kotlinx.dataframe.impl.api
+
+import org.jetbrains.kotlinx.dataframe.DataFrame
+import org.jetbrains.kotlinx.dataframe.api.RenameClause
+import org.jetbrains.kotlinx.dataframe.api.cast
+import org.jetbrains.kotlinx.dataframe.api.getColumnsWithPaths
+import org.jetbrains.kotlinx.dataframe.api.insert
+import org.jetbrains.kotlinx.dataframe.api.under
+import org.jetbrains.kotlinx.dataframe.columns.ColumnKind
+import org.jetbrains.kotlinx.dataframe.columns.ColumnWithPath
+import org.jetbrains.kotlinx.dataframe.impl.columns.tree.allChildrenNotNull
+import org.jetbrains.kotlinx.dataframe.impl.columns.tree.collectTree
+import org.jetbrains.kotlinx.dataframe.kind
+
+internal fun <T, C> RenameClause<T, C>.renameImpl(newNames: Array<out String>): DataFrame<T> {
+    var i = 0
+    return renameImpl { newNames[i++] }
+}
+
+internal fun <T, C> RenameClause<T, C>.renameImpl(transform: (ColumnWithPath<C>) -> String): DataFrame<T> {
+    val selectedColumns = df.getColumnsWithPaths(columns)
+    val tree = df.getColumnsWithPaths { all().rec() }.collectTree()
+
+    // perform rename in nodes
+    tree.allChildrenNotNull().forEach { node ->
+        val column = selectedColumns.find { it.data == node.data } ?: return@forEach
```

Comment:
What's data here? From collectTree definition it seems its a DataColumn, and in this case equals will iterate over elements of the column. Quite a heavy operation

Reply:
is it possible that there's two identical columns and second one will never be renamed?

Reply:
`data` is a column with no path anymore (`collectTree()` removes it and ColumnWithPath.data yields the original column). So to get that path back I need to find it in `selectedColumns` (I'll change this to a Map actually). I would like to use `node.pathFromRoot()` but unfortunately, the tree might already have had some node renamed, so this would generate wrong results.